lib/gitsh/completion_escaper.rb

```diff
@@ -0,0 +1,83 @@
+require 'gitsh/parser'
+
+module Gitsh
+  class CompletionEscaper
+    UNQUOTED_STRING_ESCAPABLES = Gitsh::Parser::UNQUOTED_STRING_ESCAPABLES
+    QUOTED_STRING_ESCAPABLES = {
+      '"' => Gitsh::Parser::SOFT_STRING_ESCAPABLES,
+      "'" => Gitsh::Parser::HARD_STRING_ESCAPABLES,
+    }.freeze
+
+    def initialize(completer, options)
+      @completer = completer
+      @line_editor = options[:line_editor]
+    end
+
+    def call(input)
+      OptionEscaper.new(
+        completer,
+        input,
+        line_editor.line_buffer,
+      ).escaped_options
+    end
+
+    private
+
+    attr_reader :completer, :line_editor
+
+    class OptionEscaper
+      def initialize(completer, input, full_input)
+        @completer = completer
+        @input = input
+        @full_input = full_input
+      end
+
+      def escaped_options
+        unescaped_options.map { |option| escape(option) }
+      end
+
+      private
+
+      attr_reader :completer, :input, :full_input
+
+      def unescaped_options
+        completer.call(unescape(input))
+      end
+
+      def escape(option)
+        if completing_quoted_argument?
+          quote_char = input_before_current_argument[-1]
+          escape_chars(option, QUOTED_STRING_ESCAPABLES[quote_char]).strip
+        else
+          escape_chars(option, UNQUOTED_STRING_ESCAPABLES)
+        end
+      end
+
+      def escape_chars(text, chars)
+        quotable_characters = Regexp.escape(chars)
+        text.gsub(/([#{quotable_characters}])(?!$)/) { |char| "\\#{char}" }
+      end
+
+      def unescape(input)
```

Comment:
This doesn't interpret escape sequences in the same way as the parser: we should only allow escape sequences for the characters listed in the relevant `*__STRING_ESCAPABLES` constant, any other `\` should by taken as a literal.

Basis of a test could be:

<pre><code>!touch file.txt
add f\i<kbd>tab</kbd>  # will complete to `add file.txt`, but shouldn't</code></pre>